packages/mobile/src/identity/feelessVerification.ts

```diff
@@ -0,0 +1,1061 @@
+
+function* fetchKomenciReadiness(komenciKit: KomenciKit) {
+  Logger.debug(TAG, '@fetchKomenciReadiness', 'Starting fetch')
+  const feelessVerificationState: FeelessVerificationState = yield select(
+    feelessVerificationStateSelector
+  )
+
+  const { komenci } = feelessVerificationState
+
+  if (!features.KOMENCI) {
+    throw new KomenciDisabledError()
+  }
+
+  const serviceStatusResult: Result<true, KomenciDown> = yield call([
+    komenciKit,
+    komenciKit.checkService,
+  ])
+  if (!serviceStatusResult.ok) {
+    Logger.debug(TAG, '@fetchKomenciReadiness', 'Service down')
+    throw serviceStatusResult.error
+  }
+
+  if (hasExceededKomenciErrorQuota(komenci.errorTimestamps)) {
+    Logger.debug(TAG, '@fetchKomenciReadiness', 'Too  many errors')
+    throw new KomenciErrorQuotaExceeded()
+  }
+
+  yield put(
+    feelessUpdateVerificationState({
+      ...feelessVerificationState,
+      komenci: {
+        ...feelessVerificationState.komenci,
+        serviceAvailable: true,
+      },
+    })
+  )
+}
+
+function* fetchPhoneHashDetailsFromCache(e164Number: string) {
+  Logger.debug(TAG, '@fetchPhoneHashDetailsFromCache', 'Starting fetch')
+  const [feelessVerificationState, pepperCache]: [
+    FeelessVerificationState,
+    E164NumberToSaltType
+  ] = yield all([select(feelessVerificationStateSelector), select(e164NumberToSaltSelector)])
+
+  let { phoneHashDetails } = feelessVerificationState
+  let { phoneHash } = feelessVerificationState.phoneHashDetails
+  const ownPepper = pepperCache[e164Number]
+
+  // If we already know the phoneHash, we can bail without updating state
+  if (phoneHash.length) {
+    return phoneHashDetails
+  }
+
+  // Pepper will be queried as part of feeless verification flow but cache
+  // may be empty if the user hasn't initiated a Komenci session before
+  if (!ownPepper) {
+    Logger.debug(TAG, '@fetchPhoneHashDetailsFromCache', 'Pepper not cached')
+    throw new PepperNotCachedError()
+  }
+
+  phoneHash = getPhoneHash(e164Number, ownPepper)
+  phoneHashDetails = { e164Number, phoneHash, pepper: ownPepper }
+
+  yield put(
+    feelessUpdateVerificationState({
+      ...feelessVerificationState,
+      phoneHashDetails,
+    })
+  )
+  return phoneHashDetails
+}
+
+function* fetchSessionState(komenciKit: KomenciKit, e164Number: string) {
+  Logger.debug(TAG, '@fetchSessionState', 'Starting fetch')
+  const [feelessVerificationState, pepperCache, sessionStatusResult]: [
+    FeelessVerificationState,
+    E164NumberToSaltType,
+    Result<CheckSessionResp, FetchError>
+  ] = yield all([
+    select(feelessVerificationStateSelector),
+    select(e164NumberToSaltSelector),
+    call([komenciKit, komenciKit.checkSession]),
+  ])
+
+  let sessionActive = true
+  let {
+    requestQuotaRemaining,
+    pepperQuotaRemaining,
+    unverifiedMtwAddress,
+  } = feelessVerificationState.komenci
+
+  // An inactive session is not fatal, it just means we will need to start one
+  if (!sessionStatusResult.ok) {
```

Comment:
We should check for inactive state vs other error types here.

Reply:
That would be more efficient, though any "fatal" failure that occurred here would also occur in the else block immediately following this.

@bowd what type of error gets returned when there is simply no session associated with a given `sessionToken`?

Reply:
That's true. If there is a "fatal" error in checking the state there is not much harm in trying to start a new session anyway. Worst case scenario is that the error is a timeout due to unresponsive service, in which case we either would have failed at the service readiness check or will fail to start the new session, only delaying feedback to the user, but not causing a flow to fail that would have otherwise succeeded.